pypesto/hierarchical/base_problem.py

```diff
@@ -0,0 +1,266 @@
+"""Inner optimization problem in hierarchical optimization."""
+import logging
+from typing import Dict, List, Tuple, Union
+
+import numpy as np
+import pandas as pd
+
+from .base_parameter import InnerParameter
+
+try:
+    import amici
+    import petab
+    from petab.C import OBSERVABLE_ID, TIME
+except ImportError:
+    pass
+
+logger = logging.getLogger(__name__)
+
+
+class InnerProblem:
+    """
+    Inner optimization problem in hierarchical optimization.
+
+    Attributes
+    ----------
+    xs:
+        Mapping of (inner) parameter ID to ``InnerParameters``.
+    data:
+        Measurement data. One matrix (`num_timepoints` x `num_observables`)
+        per simulation condition. Missing observations as NaN.
+    """
+
+    def __init__(self, xs: List[InnerParameter], data: List[np.ndarray]):
+        self.xs: Dict[str, InnerParameter] = {
+            x.inner_parameter_id: x for x in xs
+        }
+        self.data = data
+
+        # create the joint mask of all inner problem parameters
+        self.data_mask = [
+            np.zeros_like(cond_data, dtype=bool) for cond_data in data
+        ]
+        for x in xs:
+            for condition_ix, cond_mask in enumerate(self.data_mask):
+                cond_mask[x.ixs[condition_ix]] = True
+        # TODO should the data just be masked already here?
+        # this would make sense, since inner problems should be aware
+        # of only the data its inner parameters are relevant for...
+        # Then we wouldn't need to copy the data in the inner solver
+        # calculate_obj_function (left to discuss in PR)
```

Comment:
Yup changed it. The relative solver still needs to copy the data (as it is changing it with the `apply_offset` function), but it is a nicer implementation because the inner problems should be aware of only the data they would use.